File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/properties/playstats/playedfile/specs/GivenAConnectionThatReturnsARandomErrorCode/WhenSendingPlayedToServer.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats.playedfile.specs.GivenAConnectionThatReturnsARandomErrorCode;

import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats.playedfile.PlayedFilePlayStatsUpdater;
import com.lasthopesoftware.bluewater.shared.exceptions.HttpResponseException;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.Random;
import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenSendingPlayedToServer {

	private static Object functionEnded;
	private static HttpResponseException httpResponseException;
	private static int expectedResponseCode;

	@BeforeClass
	public static void before() throws InterruptedException, IOException {
		final Random random = new Random();
		do {
			expectedResponseCode = random.nextInt();
		} while (expectedResponseCode >= 200 && expectedResponseCode < 300);

		final IConnectionProvider connectionProvider = mock(IConnectionProvider.class);

		final HttpURLConnection urlConnection = mock(HttpURLConnection.class);
		when(urlConnection.getResponseCode()).thenReturn(expectedResponseCode);

		when(connectionProvider.getConnection("File/Played", "File=15", "FileType=Key"))
			.thenReturn(urlConnection);

		final PlayedFilePlayStatsUpdater updater = new PlayedFilePlayStatsUpdater(connectionProvider);

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		updater
			.promisePlaystatsUpdate(new ServiceFile(15))
			.then(v -> {
				functionEnded = v;

				countDownLatch.countDown();
				return null;
			})
			.excuse(e -> {
				if (e instanceof HttpResponseException) {
					httpResponseException = (HttpResponseException)e;
				}

				countDownLatch.countDown();
				return null;
			});

		countDownLatch.await();
	}

	@Test
	public void thenAnHttpResponseExceptionIsThrownWithTheResponseCode() {
		assertThat(httpResponseException.getResponseCode()).isEqualTo(expectedResponseCode);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/files/properties/playstats/playedfile/GivenAConnectionThatReturnsA404ErrorCode/WhenSendingPlayedToServer.kt
package com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats.playedfile.GivenAConnectionThatReturnsA404ErrorCode

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats.playedfile.PlayedFilePlayStatsUpdater
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionProvider
import com.lasthopesoftware.bluewater.shared.exceptions.HttpResponseException
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.promises.extensions.ProgressingPromise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.concurrent.ExecutionException

class WhenSendingPlayedToServer {

	private var httpResponseException: HttpResponseException? = null

	@BeforeAll
	fun act() {
		val updater = PlayedFilePlayStatsUpdater(
			mockk {
				every { promiseLibraryConnection(any()) } returns ProgressingPromise(FakeConnectionProvider())
			}
		)
		try {
			updater.promisePlaystatsUpdate(LibraryId(394), ServiceFile(15)).toExpiringFuture().get()
		} catch (e: ExecutionException) {
			httpResponseException = e.cause as? HttpResponseException?
		}
	}

	@Test
	fun thenAnHttpResponseExceptionIsThrownWithTheResponseCode() {
		assertThat(httpResponseException!!.responseCode).isEqualTo(404)
	}
}
